TypeCobol.Analysis/Graph/CfgAbstractInterpretation.Environment.cs

```diff
@@ -0,0 +1,341 @@
+        /// branching instructions are handled specifically.
+        /// Thus not all execution paths are applied, but all nodes are visited using dfs like algorithm if they are accessible.
+        /// A node cannot be visited twice.
+        /// </summary>
+        public class Environment
+        {
+            /// <summary>
+            /// The Abstract Interpretation Stack
+            /// </summary>
+            private Stack<BasicBlock<N, D>> _interpretationStack;        
+
+            /// <summary>
+            /// The current execution stack
+            /// </summary>
+            public IReadOnlyCollection<BasicBlock<N, D>> ExecutionStack => _interpretationStack;
+
+            /// <summary>
+            /// The DFS Run stack
+            /// </summary>
+            private Stack<BasicBlock<N, D>> _dfsStack;            
+            /// <summary>
+            /// Bit Vector to check if metric values have been taken in account for a block index.
+            /// </summary>
+            private System.Collections.BitArray _metricUpdated;
+            /// <summary>
+            /// Current Cyclic execution Thresold by block index.
+            /// </summary>
+            private Dictionary<int,int> _cyclicThreshold;
+            /// <summary>
+            /// The Cyclic relation from one block index to another block index
+            /// </summary>
+            private Dictionary<int, HashSet<int> >_cyclicRelation;
+            /// <summary>
+            /// Number maximal of cyclic execution for a block index.
+            /// </summary>
+            private int _cyclicExecutionThresold ;
+            /// <summary>
+            /// Current CFG being run.
+            /// </summary>
+            public ControlFlowGraph<N, D> Cfg { get; private set; }
+            private readonly IObserver[] _observers;
+            /// <summary>
+            /// Some metrics calculated.
+            /// </summary>
+            private Metrics _metrics;
+            public Environment(params IObserver[] observers)
+            {
+                this._observers = observers;
+            }
+
+            /// <summary>
+            /// Intialize internal data structures.
+            /// </summary>
+            private void Initialize()
+            {
+                if (_interpretationStack == null)
+                    _interpretationStack = new Stack<BasicBlock<N, D>>();
+                if (_dfsStack == null)
+                    _dfsStack = new Stack<BasicBlock<N, D>>();
+                _metricUpdated = new System.Collections.BitArray(Cfg.AllBlocks.Count);
+                if (_cyclicThreshold == null)
+                    _cyclicThreshold = new Dictionary<int, int>();
+                else
+                    _cyclicThreshold.Clear();
+                if (_cyclicRelation == null)
+                    _cyclicRelation = new Dictionary<int, HashSet<int>>();
+                else
+                    _cyclicRelation.Clear();
```

Comment:
It is renamed Reset, no need to have it in the constructor, beacuse it is any call to Run that need to Reset value.

Reply:
doen in 7c56456

Reply:
Init operations in constructor and Clear operations in Reset, this would avoird testing for null at each run. This was my idea but ok to keep it like this.